In [7]:
import pandas as pd

from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [2]:
!pip -q install optuna

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 413.9/413.9 kB 33.0 MB/s eta 0:00:00


In [8]:
import optuna
from optuna.pruners import MedianPruner # הוספת הגיזום
from transformers import set_seed
import numpy as np
import itertools
from datasets import load_dataset
from transformers import (
    AutoTokenizer,
    AutoModelForSequenceClassification,
    DataCollatorWithPadding,
    TrainingArguments,
    Trainer,
    EarlyStoppingCallback,
)
from sklearn.metrics import f1_score, accuracy_score

# -----------------------
# A) Load JSONL
# -----------------------
data_files = {
    "train": "/content/drive/MyDrive/nlp_pro2/data_sets/use_part/train.jsonl",
    "validation": "/content/drive/MyDrive/nlp_pro2/data_sets/use_part/val.jsonl",
    "test": "/content/drive/MyDrive/nlp_pro2/data_sets/use_part/test.jsonl",
}
ds = load_dataset("json", data_files=data_files)

# -----------------------
# B) Label mapping
# -----------------------
label_list = ["entailment", "contradiction", "neutral"]
label2id = {l: i for i, l in enumerate(label_list)}
id2label = {i: l for l, i in label2id.items()}

def add_label_id(example):
    example["label"] = label2id[str(example["label"]).lower()]
    return example

ds = ds.map(add_label_id)

# -----------------------
# C) Tokenizer
# -----------------------
model_name = "onlplab/alephbert-base"
tokenizer = AutoTokenizer.from_pretrained(model_name)

if tokenizer.pad_token is None:
    tokenizer.add_special_tokens({"pad_token": "[PAD]"})

def tokenize_pair(example):
    return tokenizer(
        example["translation1"],
        example["translation2"],
        truncation=True,
        max_length=128,
    )

ds_tok = ds.map(tokenize_pair, batched=False)
cols = ["input_ids", "attention_mask", "label"]
ds_tok = ds_tok.remove_columns([c for c in ds_tok["train"].column_names if c not in cols])

data_collator = DataCollatorWithPadding(tokenizer=tokenizer)

# -----------------------
# D) Metrics
# -----------------------
def compute_metrics(eval_pred):
    logits, labels = eval_pred
    preds = np.argmax(logits, axis=-1)
    return {
        "accuracy": accuracy_score(labels, preds),
        "macro_f1": f1_score(labels, preds, average="macro"),
    }

# -----------------------
# E) Model init (fresh model per run)
# -----------------------
def model_init():
    m = AutoModelForSequenceClassification.from_pretrained(
        model_name,
        num_labels=3,
        id2label=id2label,
        label2id=label2id,
    )
    # if we added pad token, keep embeddings consistent
    if len(tokenizer) != m.config.vocab_size:
        m.resize_token_embeddings(len(tokenizer))
    return m


set_seed(42)

def hp_space(trial: optuna.Trial):
    return {
        "learning_rate": trial.suggest_float("learning_rate", 2e-5, 5e-5, log=True),
        "weight_decay": trial.suggest_float("weight_decay", 0.01, 0.1),
        "warmup_ratio": trial.suggest_float("warmup_ratio", 0.0, 0.1),
        "num_train_epochs": trial.suggest_int("num_train_epochs", 2, 3),
        "per_device_train_batch_size": 32, # קיבוע ל-32 לחיסכון בזמן (A100 מתמודד עם זה בקלות)
    }

def compute_objective(metrics):
    # HF will pass eval metrics dict here
    return metrics["eval_macro_f1"]

# --- עדכון ה-TrainingArguments הבסיסיים ---
base_args = TrainingArguments(
    output_dir="hebert_optuna",
    eval_strategy="epoch",
    save_strategy="epoch",
    logging_strategy="epoch",
    report_to="none",
    seed=42,
    fp16=True, # האצה משמעותית ב-A100
    load_best_model_at_end=True,
    metric_for_best_model="macro_f1",
    greater_is_better=True,
    per_device_eval_batch_size=32, # Batch קבוע להערכה מהירה
)

trainer = Trainer(
    args=base_args,
    model_init=model_init,
    train_dataset=ds_tok["train"],
    eval_dataset=ds_tok["validation"],
    data_collator=data_collator,
    compute_metrics=compute_metrics,
    callbacks=[EarlyStoppingCallback(early_stopping_patience=1)],
)


# הפעלת החיפוש עם Pruner
best_run = trainer.hyperparameter_search(
    backend="optuna",
    direction="maximize",
    hp_space=hp_space,
    compute_objective=compute_objective,
    n_trials=10, # 10 ניסויים מספיקים לרוב עם Bayesian Optimization
    pruner=MedianPruner(), # עוצר ניסויים חלשים באמצע
)

print("Best trial:", best_run)
best_cfg = best_run.hyperparameters
print("BEST CFG:", best_cfg)


# -----------------------
# G) FINAL TRAIN with best hyperparameters (then test ONCE)
# -----------------------
final_args = TrainingArguments(
    output_dir="hebert_final",
    eval_strategy="epoch",
    save_strategy="epoch",
    load_best_model_at_end=True,
    metric_for_best_model="macro_f1",
    greater_is_better=True,
    learning_rate=best_cfg["learning_rate"],
    weight_decay=best_cfg["weight_decay"],
    warmup_ratio=best_cfg["warmup_ratio"],
    per_device_train_batch_size=best_cfg["per_device_train_batch_size"],
    per_device_eval_batch_size=32,
    num_train_epochs=best_cfg["num_train_epochs"],
    report_to="none",
    seed=42,
)

final_trainer = Trainer(
    args=final_args,
    model_init=model_init,
    train_dataset=ds_tok["train"],
    eval_dataset=ds_tok["validation"],
    data_collator=data_collator,
    compute_metrics=compute_metrics,
)

final_trainer.train()

print("\nTEST:")
test_metrics = final_trainer.evaluate(ds_tok["test"])
print(test_metrics)


Generating train split: 0 examples [00:00, ? examples/s]

Generating validation split: 0 examples [00:00, ? examples/s]

Generating test split: 0 examples [00:00, ? examples/s]

Map:   0%|          | 0/99999 [00:00<?, ? examples/s]

Map:   0%|          | 0/21429 [00:00<?, ? examples/s]

Map:   0%|          | 0/21429 [00:00<?, ? examples/s]

Map:   0%|          | 0/99999 [00:00<?, ? examples/s]

Map:   0%|          | 0/21429 [00:00<?, ? examples/s]

Map:   0%|          | 0/21429 [00:00<?, ? examples/s]

Some weights of BertForSequenceClassification were not initialized from the model checkpoint at onlplab/alephbert-base and are newly initialized: ['bert.pooler.dense.bias', 'bert.pooler.dense.weight', 'classifier.bias', 'classifier.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.
[I 2026-01-31 23:26:36,398] A new study created in memory with name: no-name-2bb814ef-6fee-467b-8980-de54c38c4977
Some weights of BertForSequenceClassification were not initialized from the model checkpoint at onlplab/alephbert-base and are newly initialized: ['bert.pooler.dense.bias', 'bert.pooler.dense.weight', 'classifier.bias', 'classifier.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


Epoch,Training Loss,Validation Loss,Accuracy,Macro F1
1,0.803600,0.691997,0.711886,0.710421
2,0.521500,0.697245,0.732045,0.731744


[I 2026-01-31 23:32:16,987] Trial 0 finished with value: 0.7317442453619364 and parameters: {'learning_rate': 3.751206772224478e-05, 'weight_decay': 0.07437065906234097, 'warmup_ratio': 0.060951322616641074, 'num_train_epochs': 2}. Best is trial 0 with value: 0.7317442453619364.
Some weights of BertForSequenceClassification were not initialized from the model checkpoint at onlplab/alephbert-base and are newly initialized: ['bert.pooler.dense.bias', 'bert.pooler.dense.weight', 'classifier.bias', 'classifier.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


Epoch,Training Loss,Validation Loss,Accuracy,Macro F1
1,0.804600,0.694944,0.709039,0.706973
2,0.523700,0.692695,0.729105,0.728800


[I 2026-01-31 23:37:55,348] Trial 1 finished with value: 0.7288000102115446 and parameters: {'learning_rate': 3.6207333945789344e-05, 'weight_decay': 0.05776802815054454, 'warmup_ratio': 0.06858476359562964, 'num_train_epochs': 2}. Best is trial 0 with value: 0.7317442453619364.
Some weights of BertForSequenceClassification were not initialized from the model checkpoint at onlplab/alephbert-base and are newly initialized: ['bert.pooler.dense.bias', 'bert.pooler.dense.weight', 'classifier.bias', 'classifier.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


Epoch,Training Loss,Validation Loss,Accuracy,Macro F1
1,0.794000,0.692117,0.712166,0.710484
2,0.546100,0.696382,0.729525,0.728983
3,0.358100,0.814524,0.727472,0.727114


[I 2026-01-31 23:46:22,331] Trial 2 finished with value: 0.7271135253927058 and parameters: {'learning_rate': 2.7259114538937245e-05, 'weight_decay': 0.09658807955509473, 'warmup_ratio': 0.014544503121575415, 'num_train_epochs': 3}. Best is trial 0 with value: 0.7317442453619364.
Some weights of BertForSequenceClassification were not initialized from the model checkpoint at onlplab/alephbert-base and are newly initialized: ['bert.pooler.dense.bias', 'bert.pooler.dense.weight', 'classifier.bias', 'classifier.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


Epoch,Training Loss,Validation Loss,Accuracy,Macro F1
1,0.790800,0.691342,0.712866,0.711293
2,0.563800,0.689796,0.729945,0.729404
3,0.402500,0.774557,0.728125,0.727756


[I 2026-01-31 23:54:48,479] Trial 3 finished with value: 0.7277561299386147 and parameters: {'learning_rate': 2.0997442359613833e-05, 'weight_decay': 0.03215509261337169, 'warmup_ratio': 0.0040966442219941505, 'num_train_epochs': 3}. Best is trial 0 with value: 0.7317442453619364.
Some weights of BertForSequenceClassification were not initialized from the model checkpoint at onlplab/alephbert-base and are newly initialized: ['bert.pooler.dense.bias', 'bert.pooler.dense.weight', 'classifier.bias', 'classifier.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


Epoch,Training Loss,Validation Loss,Accuracy,Macro F1
1,0.811500,0.685085,0.714966,0.713688
2,0.525600,0.687103,0.733679,0.733426


[I 2026-02-01 00:00:29,846] Trial 4 finished with value: 0.733426443659693 and parameters: {'learning_rate': 3.8974309591966866e-05, 'weight_decay': 0.026734814827107628, 'warmup_ratio': 0.09792891307025368, 'num_train_epochs': 2}. Best is trial 4 with value: 0.733426443659693.
Some weights of BertForSequenceClassification were not initialized from the model checkpoint at onlplab/alephbert-base and are newly initialized: ['bert.pooler.dense.bias', 'bert.pooler.dense.weight', 'classifier.bias', 'classifier.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


Epoch,Training Loss,Validation Loss,Accuracy,Macro F1
1,0.821600,0.691259,0.710299,0.709245


[I 2026-02-01 00:03:17,245] Trial 5 pruned. 
Some weights of BertForSequenceClassification were not initialized from the model checkpoint at onlplab/alephbert-base and are newly initialized: ['bert.pooler.dense.bias', 'bert.pooler.dense.weight', 'classifier.bias', 'classifier.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


Epoch,Training Loss,Validation Loss,Accuracy,Macro F1
1,0.786400,0.685943,0.711699,0.710176


[I 2026-02-01 00:06:05,971] Trial 6 pruned. 
Some weights of BertForSequenceClassification were not initialized from the model checkpoint at onlplab/alephbert-base and are newly initialized: ['bert.pooler.dense.bias', 'bert.pooler.dense.weight', 'classifier.bias', 'classifier.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


Epoch,Training Loss,Validation Loss,Accuracy,Macro F1
1,0.821700,0.696658,0.707546,0.706038


[I 2026-02-01 00:08:54,460] Trial 7 pruned. 
Some weights of BertForSequenceClassification were not initialized from the model checkpoint at onlplab/alephbert-base and are newly initialized: ['bert.pooler.dense.bias', 'bert.pooler.dense.weight', 'classifier.bias', 'classifier.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


Epoch,Training Loss,Validation Loss,Accuracy,Macro F1
1,0.807800,0.689635,0.711512,0.710060


[I 2026-02-01 00:11:43,100] Trial 8 pruned. 
Some weights of BertForSequenceClassification were not initialized from the model checkpoint at onlplab/alephbert-base and are newly initialized: ['bert.pooler.dense.bias', 'bert.pooler.dense.weight', 'classifier.bias', 'classifier.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


Epoch,Training Loss,Validation Loss,Accuracy,Macro F1
1,0.811700,0.690329,0.712492,0.710839
2,0.527600,0.689087,0.733912,0.733691


[I 2026-02-01 00:17:24,102] Trial 9 finished with value: 0.7336907672546618 and parameters: {'learning_rate': 3.76690649380035e-05, 'weight_decay': 0.039841378722297226, 'warmup_ratio': 0.09881535889557601, 'num_train_epochs': 2}. Best is trial 9 with value: 0.7336907672546618.


Best trial: BestRun(run_id='9', objective=0.7336907672546618, hyperparameters={'learning_rate': 3.76690649380035e-05, 'weight_decay': 0.039841378722297226, 'warmup_ratio': 0.09881535889557601, 'num_train_epochs': 2}, run_summary=None)
BEST CFG: {'learning_rate': 3.76690649380035e-05, 'weight_decay': 0.039841378722297226, 'warmup_ratio': 0.09881535889557601, 'num_train_epochs': 2}


KeyError: 'per_device_train_batch_size'

BEST CFG:
*  'learning_rate': 3.76690649380035e-05,
* 'weight_decay': 0.039841378722297226,
* 'warmup_ratio': 0.09881535889557601,